converter/interceptor.go

```diff
@@ -0,0 +1,215 @@
+		case enumspb.COMMAND_TYPE_SCHEDULE_ACTIVITY_TASK:
+			err = encodePayloads(c.GetScheduleActivityTaskCommandAttributes().Input, s.encoders...)
+		case enumspb.COMMAND_TYPE_COMPLETE_WORKFLOW_EXECUTION:
+			err = encodePayloads(c.GetCompleteWorkflowExecutionCommandAttributes().Result, s.encoders...)
+		case enumspb.COMMAND_TYPE_CONTINUE_AS_NEW_WORKFLOW_EXECUTION:
+			err = encodePayloads(c.GetContinueAsNewWorkflowExecutionCommandAttributes().Input, s.encoders...)
+		case enumspb.COMMAND_TYPE_START_CHILD_WORKFLOW_EXECUTION:
+			err = encodePayloads(c.GetStartChildWorkflowExecutionCommandAttributes().Input, s.encoders...)
+		case enumspb.COMMAND_TYPE_SIGNAL_EXTERNAL_WORKFLOW_EXECUTION:
+			err = encodePayloads(c.GetSignalExternalWorkflowExecutionCommandAttributes().Input, s.encoders...)
+		}
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *serviceInterceptor) processEvents(events []*historypb.HistoryEvent) error {
+	var err error
+
+	for _, e := range events {
+		switch e.EventType {
+		case enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_STARTED:
+			err = decodePayloads(e.GetWorkflowExecutionStartedEventAttributes().Input, s.encoders...)
+		case enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_COMPLETED:
+			err = decodePayloads(e.GetWorkflowExecutionCompletedEventAttributes().Result, s.encoders...)
+		case enumspb.EVENT_TYPE_START_CHILD_WORKFLOW_EXECUTION_INITIATED:
+			err = decodePayloads(e.GetStartChildWorkflowExecutionInitiatedEventAttributes().Input, s.encoders...)
+		case enumspb.EVENT_TYPE_SIGNAL_EXTERNAL_WORKFLOW_EXECUTION_INITIATED:
+			err = decodePayloads(e.GetSignalExternalWorkflowExecutionInitiatedEventAttributes().Input, s.encoders...)
+		case enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_CONTINUED_AS_NEW:
+			err = decodePayloads(e.GetWorkflowExecutionContinuedAsNewEventAttributes().Input, s.encoders...)
+		case enumspb.EVENT_TYPE_ACTIVITY_TASK_SCHEDULED:
+			err = decodePayloads(e.GetActivityTaskScheduledEventAttributes().Input, s.encoders...)
+		case enumspb.EVENT_TYPE_ACTIVITY_TASK_COMPLETED:
+			err = decodePayloads(e.GetActivityTaskCompletedEventAttributes().Result, s.encoders...)
+		case enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_SIGNALED:
+			err = decodePayloads(e.GetWorkflowExecutionSignaledEventAttributes().Input, s.encoders...)
+		case enumspb.EVENT_TYPE_CHILD_WORKFLOW_EXECUTION_COMPLETED:
+			err = decodePayloads(e.GetChildWorkflowExecutionCompletedEventAttributes().Result, s.encoders...)
+		}
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// NewPayloadEncoderGRPCServerInterceptor returns a GRPC Server Interceptor that will mimic the encoding
+// that the SDK system would perform when configured with a matching EncodingDataConverter.
+// Note: This approach does not support use cases that rely on the ContextAware DataConverter interface as
+// workflow context is not available at the GRPC level.
+func NewPayloadEncoderGRPCServerInterceptor(encoders ...PayloadEncoder) grpc.UnaryServerInterceptor {
```

Comment:
What is the use case for an encoding server interceptor?

Reply:
It feels as relevant as the client side one to me. For the proxy use case we can use either but I don't know if others would need server specifically.

Reply:
Server makes less sense in my head since I don't expect many people to run a server, but it's pretty cheap to add here, so I don't mind.